sqlglot/dataframe/sql/dataframe.py

```diff
@@ -0,0 +1,737 @@
+        expression_to_cte.set("with", None)
+        cte = exp.Select().with_(name, as_=expression_to_cte, **kwargs).ctes[0]
+        cte.set("branch_id", branch_id or self.branch_id)
+        cte.set("sequence_id", sequence_id or self.sequence_id)
+        return cte, name
+
+    def _ensure_list_of_columns(
+        self, cols: t.Union[str, t.Iterable[str], Column, t.Iterable[Column]]
+    ) -> t.List[Column]:
+        columns = ensure_list(cols)
+        columns = Column.ensure_cols(columns)
+        return columns
+
+    def _ensure_and_normalize_cols(self, cols):
+        cols = self._ensure_list_of_columns(cols)
+        normalize(self.spark, self.expression, cols)
+        return cols
+
+    def _ensure_and_normalize_col(self, col):
+        col = Column.ensure_col(col)
+        normalize(self.spark, self.expression, col)
+        return col
+
+    def _convert_leaf_to_cte(self, sequence_id: t.Optional[str] = None) -> DataFrame:
+        df = self._resolve_pending_hints()
+        sequence_id = sequence_id or df.sequence_id
+        expression = df.expression.copy()
+        cte_expression, cte_name = df._create_cte_from_expression(expression=expression, sequence_id=sequence_id)
+        new_expression = exp.Select()
+        new_expression = df._add_ctes_to_expression(new_expression, expression.ctes + [cte_expression])
+        sel_columns = df._get_outer_select_columns(cte_expression)
+        new_expression = new_expression.from_(cte_name).select(*[x.alias_or_name for x in sel_columns])
+        return df.copy(expression=new_expression, sequence_id=sequence_id)
+
+    def _resolve_pending_hints(self) -> DataFrame:
+        df = self.copy()
+        if not self.pending_hints:
+            return df
+        expression = df.expression
+        hint_expression = expression.args.get("hint") or exp.Hint(expressions=[])
+        for hint in df.pending_partition_hints:
+            hint_expression.args.get("expressions").append(hint)
+            df.pending_hints.remove(hint)
+
+        join_tables = get_tables_from_expression_with_join(expression)
+        if join_tables:
+            for hint in df.pending_join_hints:
+                for sequence_id_expression in hint.expressions:
+                    sequence_id_or_name = sequence_id_expression.alias_or_name
+                    sequence_ids_to_match = [sequence_id_or_name]
+                    if sequence_id_or_name in df.spark.name_to_sequence_id_mapping:
+                        sequence_ids_to_match = df.spark.name_to_sequence_id_mapping[sequence_id_or_name]
+                    matching_ctes = [
+                        cte for cte in reversed(expression.ctes) if cte.args["sequence_id"] in sequence_ids_to_match
+                    ]
+                    if matching_ctes:
+                        for matching_cte in matching_ctes:
+                            if matching_cte.alias_or_name in [join_table.alias_or_name for join_table in join_tables]:
```

Comment:
take this for loop outside and make it a set?

join_aliases = {join_table.alias_or_name for . in .}